include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,356 @@
+/**
+ * @file moead_impl.hpp
+ * @author Utkarsh Rai
+ *
+ * Implementation of the MOEA/D algorithm. Used for multi-objective
+ * optimization problems on arbitrary functions.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more Information.
+ */
+
+#ifndef ENSMALLEN_MOEAD_MOEAD_IMPL_HPP
+#define ENSMALLEN_MOEAD_MOEAD_IMPL_HPP
+
+#include "moead.hpp"
+
+namespace ens {
+
+inline MOEAD::MOEAD(const size_t populationSize,
+                    const double crossoverProb,
+                    const double mutationProb,
+                    const double mutationStrength,
+                    const size_t neighbourhoodSize,
+                    const arma::vec& lowerBound,
+                    const arma::vec& upperBound) :
+    populationSize(populationSize),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    neighbourhoodSize(neighbourhoodSize),
+    lowerBound(lowerBound),
+    upperBound(upperBound),
+    numObjectives(0)
+  { /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename MatType,
+         typename... ArbitraryFunctionType,
+         typename... CallbackTypes>
+typename MatType::elem_type MOEAD::Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+                                            MatType& iterate,
+                                            CallbackTypes&&... callbacks)
+{
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+
+  if(lowerBound.size() != upperBound.size())
+  {
+    throw std::logic_error("MOEAD::Optimize(): size of lowerBound and upperBound "
+        "must be equal.");
+  }
+  if(lowerBound.size() != iterate.n_elem)
+  {
+    throw std::logic_error("MOEAD::Optimize(): there should be a lower bound and "
+        "an upper bound for each variable in the initial point.");
+  }
+  // Number of objective functions. Represented by m in the paper.
+  numObjectives = sizeof...(ArbitraryFunctionType);
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // 1.1 The external population, non-dominated solutions.
+  std::vector<MatType> externalPopulation;
+  std::vector<arma::vec> externalPopulationFValue;
+
+  // Weight vectors, where each one of them represents a decomposition.
+  std::vector<arma::vec> weights(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+    weights[i] = arma::vec(numObjectives, arma::fill::randu);
+
+  // 1.2 Storing the indices of nearest neighbours of each weight vector.
+  arma::Mat<size_t> B(populationSize, neighbourhoodSize);
```

Comment:
We could perhaps rename `B` to something more descriptive like, say, `indices`. Also, if you are unsure about the naming conventions, you can read them [here](https://github.com/mlpack/mlpack/wiki/DesignGuidelines#naming-conventions)! :)

Reply:
Thanks for reviewing it so quickly. Actually I used `B` because it was used in the paper. However, now I have changed it to `weightNeighbourIndices` and also fixed the indentation problems. Also, what might be the reason behind appveyor failing? I tried to read the console output but couldn't figure anything out.

Reply:
Ah, yes I thought so too! We should go with descriptive names whenever possible. :)

Reply:
About the AppVeyor build, I see [C4514](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-4-c4514?view=vs-2019).

Reply:
I am still unable to figure out the problem. Can you please elaborate. I see that many functions in moead give C4514 but there are other optimisers giving the same warning. Am I missing something?

Reply:
Hi @say4n, when I run `./ensmallen_test -d yes`, `MOEADSchafferN1Test` fails (The front contains only one value -0.0900101). But when I run `./ensmallen [MOEADTest] -d yes`, `MOEADSchafferN1Test` passes (The front has many values ranging from 0.45 to 0.53). Maybe this is the reason why appveyor build is failing because on comparing the console output in this PR to that in #211, I found it is exactly the same till the point the tests start running. While implementing the code I never ran `./ensmallen_test` (because it takes some time) only checked my code with `./ensmallen_test [MOEADTest]`, because I did not know this could happen. What might be the reason?

Reply:
I went through the AppVeyor logs again, at a glance, I don't see the `ctest` command (which runs the unit tests) being executed. Let me get back to you in a bit.

Reply:
Ah, from what I can tell, it looks like your build is not reaching the test stage. It is terminating before executing [these](https://github.com/mlpack/ensmallen/blob/c73c92a2f57fa256d2db50b314731635147bdd9d/.appveyor.yml#L55). Now, as to why that is happening, I have no clue! 😰

Reply:
What you could do is set up AppVeyor on your own fork and enter the build VM to triage the issue. Instructions [here](https://www.appveyor.com/docs/how-to/ssh-to-build-worker/). :)

Reply:
> Hi @say4n, when I run `./ensmallen_test -d yes`, `MOEADSchafferN1Test` fails (The front contains only one value -0.0900101). But when I run `./ensmallen [MOEADTest] -d yes`, `MOEADSchafferN1Test` passes (The front has many values ranging from 0.45 to 0.53). Maybe this is the reason why appveyor build is failing because on comparing the console output in this PR to that in #211, I found it is exactly the same till the point the tests start running. While implementing the code I never ran `./ensmallen_test` (because it takes some time) only checked my code with `./ensmallen_test [MOEADTest]`, because I did not know this could happen. What might be the reason?

And what might be the reason behind this?

Reply:
Ah, I'll have to see. :)

Reply:
Hi @say4n, sorry for reaching back so late. I will try to figure out what happened by accessing the build VM. Since I have not done this before, I have a question. The docs you shared say 
> Add the following Bash command at the place where SSH access should be enabled:
> `curl -sflL 'https://raw.githubusercontent.com/appveyor/ci/master/scripts/enable-ssh.sh' | bash -e -`

Should I add this line just before the build seems to terminate or at the very beginning just after build_script?

Reply:
Will this work as expected?
```
environment:
  ARMADILLO_DOWNLOAD: "http://ftp.fau.de/macports/distfiles/armadillo/armadillo-8.400.0.tar.xz"
  BLAS_LIBRARY: "%APPVEYOR_BUILD_FOLDER%/OpenBLAS.0.2.14.1/lib/native/lib/x64/libopenblas.dll.a"
  BLAS_LIBRARY_DLL: "%APPVEYOR_BUILD_FOLDER%/OpenBLAS.0.2.14.1/lib/native/lib/x64/libopenblas.dll"
  APPVEYOR_SSH_KEY: <your ssh public key>

  matrix:
    - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2015
      VSVER: Visual Studio 14 2015 Win64
      MSBUILD: C:\Program Files (x86)\MSBuild\14.0\bin\MSBuild.exe

    - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2017
      VSVER: Visual Studio 15 2017 Win64
      MSBUILD: C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\MSBuild.exe

    - APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2019
      VSVER: Visual Studio 16 2019
      MSBUILD: C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\MSBuild.exe

configuration: Release

install:
  - ps: nuget install OpenBLAS -o "${env:APPVEYOR_BUILD_FOLDER}"

build_script:
  # First, download and build Armadillo.
  - cd ..
  - appveyor DownloadFile %ARMADILLO_DOWNLOAD% -FileName armadillo.tar.xz
  - 7z x armadillo.tar.xz -so -txz | 7z x -si -ttar > nul
  - cd armadillo-8.400.0 && mkdir build && cd build
  - >
    cmake -G "%VSVER%"
    -DBLAS_LIBRARY:FILEPATH=%BLAS_LIBRARY%
    -DLAPACK_LIBRARY:FILEPATH=%BLAS_LIBRARY%
    -DCMAKE_PREFIX:FILEPATH="%APPVEYOR_BUILD_FOLDER%/armadillo"
    -DBUILD_SHARED_LIBS=OFF
    -DCMAKE_BUILD_TYPE=Release ..
  - >
    "%MSBUILD%" "armadillo.sln"
    /m /verbosity:quiet /p:Configuration=Release;Platform=x64
  - cd ../..

  # Now build ensmallen.
  - cd ensmallen && mkdir build && cd build
  - >
    cmake -G "%VSVER%"
    -DARMADILLO_INCLUDE_DIR=%APPVEYOR_BUILD_FOLDER%/../armadillo-8.400.0/include/
    -DARMADILLO_LIBRARIES=%BLAS_LIBRARY%
    -DLAPACK_LIBRARY=%BLAS_LIBRARY%
    -DBLAS_LIBRARY=%BLAS_LIBRARY%
    -DCMAKE_BUILD_TYPE=Release ..
  - >
    "%MSBUILD%" "ensmallen.sln"
    /m /verbosity:minimal /nologo /p:BuildInParallel=true
  - sh: curl -sflL 'https://raw.githubusercontent.com/appveyor/ci/master/scripts/enable-ssh.sh' | bash -e -
  # Run tests after copying libraries.
  - ps: cp C:\projects\ensmallen\OpenBLAS.0.2.14.1\lib\native\bin\x64\*.* C:\projects\ensmallen\build\
  - ctest -C Release -V --output-on-failure .
```

Reply:
You should ideally set it to the place just before the failing command, try and experiment with it. You can even ssh in the `init` phase and follow the logs locally and break in when required.

You also have the `APPVEYOR_SSH_BLOCK` flag, you can use it too to pause the build till you have succefully ssh-ed and the continue by deleting the lock file.

Reply:
So the first of the three builds failed. How will I be able to access the VM, I can't seem to figure it on my own. Do I need some virtualization software or a Windows machine?

Reply:
Ah, the build should pause when you set the `APPVEYOR_SSH_BLOCK` to true. It should only resume once you've ssh-ed into the machine and you've deleted the lock file that is created as per the docs [here](https://www.appveyor.com/docs/how-to/ssh-to-build-worker/). For linux you can just type in ssh `<username>@<ip-address>` to ssh to the server, on Windows, you can use something like putty.

Reply:
Glad you figured it out!
(although, wrong username tag 😛)

Reply:
Locally the tests pass but they are still failing (on two builds now).

Reply:
Hm, you could try tweaking the hyper-parameters. :)

Reply:
The hyper-parameter values I had set (N=1000, T=200) were very different from the ones recommended in the paper (N=150, T=10). Let's hope it passes this time.

Reply:
First of the three appveyor builds passed in my repository!! Hopefully it passes here too. There are some undocumented variables. I'll make the changes once(if) the builds pass here.

Reply:
The first appveyor builld here failed because of NSGA2Test and not MOEADTest. Can you please take a look?